GVRf/Framework/src/org/gearvrf/GVRHybridObject.java

```diff
@@ -15,209 +15,97 @@
-     */
-    private final GVRReference mReference;
+    private long mNativePointer;
 
-    private int mFlags = sGenerationCounter;
-
-    GVRHybridObject(GVRContext gvrContext, long ptr) {
+    GVRHybridObject(GVRContext gvrContext, long nativePointer) {
         mGVRContext = gvrContext;
-
-        GVRReferenceQueue referenceQueue = gvrContext.getReferenceQueue();
-        if (this instanceof GVRRecyclableObject) {
-            mReference = new GVRRecyclableReference(ptr,
-                    (GVRRecyclableObject) this,
-                    referenceQueue.getRecyclableQueue());
-        } else {
-            mReference = new GVRHybridReference(ptr, this,
-                    referenceQueue.getHybridReferenceQueue());
-        }
-
-        /*
-         * Needed to save the reference from being garbage collected before the
-         * linked hybrid object gets collected.
-         */
-        referenceQueue.addReference(mReference);
-
-        if (registerWrapper()) {
-            // 'Register' fully initialized object, so that getX() calls can
-            // return this (possibly sub-classed) object, not a new wrapper
-            final long nativePointer = NativeHybridObject.getNativePointer(ptr);
-            synchronized (sWrappers) {
-                sWrappers.put(nativePointer, this);
-
-                if ((++sRegistrationCount % REGISTRATIONS_BETWEEN_DEREFERENCE_SCANS) == 0) {
-                    sRegistrationCount = 0;
-                    synchronized (sDereferenceThread) {
-                        sDereferenceThread.notify();
-                    }
-                }
-            }
-        }
+        mNativePointer = nativePointer;
     }
 
     /**
-     * Set or clear the keep-wrapper flag.
+     * The {@code finalize} method is {@code final} to protect the
+     * {@link #close()} 'machinery' in {@link GVRHybridObject}.
      * 
-     * This is a very specialized operation: you will not use it often. The only
-     * consequences of misuse are a small memory leak, but you should still know
-     * what you're doing before calling this method.
+     * If you need to run any code when a descendant is closed, override the
+     * {@link #destructor()} method.
      * 
-     * <p>
-     * You only need to set the keep-wrapper flag <em>in some cases</em> when
-     * you're getting a GVRF object from another GVRF object, and it really
-     * matters that the object that you get back is the object you put in (as
-     * opposed to another object that refers to the same native object). The
-     * most common scenario where this matters is one where you have declared
-     * classes that {@code extend GVRSceneObject} and you want to get your
-     * sub-classed object back from the {@link GVRPicker} because that instance
-     * contains app-specific values that a 'generic' {@link GVRSceneObject} does
-     * not.
+     * @since 2.0.0
+     */
+    @Override
+    protected final void finalize() throws Throwable {
```

Comment:
We've had some internal (off GitHub) discussion about the costs of using `finalize()`. While it looked like I _had_ to do this (because (at least on Lollipop) `Reference.get()` always returns `null` once the `Reference` has been placed on a `ReferenceQueue`, for all three `Reference` types) @ejsamsung suggested the clever work-around of using a class derived from `PhantomReference` that has a copy of the native pointer, so it can call `NativeHybridObject.delete()`.

The only problem with this optimization is that by the time a `Reference` has been placed on a `ReferenceQueue`, it has already been finalized: `~HybridObject()` _can't_ call an `onDestroy()` method that would (say) unregister `GVREyePointeeHolder`. I mention this now as a sort of early warning: the `destructor()` method is probably going to disappear in the very near future.